sw/device/lib/pinmux.c

```diff
@@ -6,36 +6,40 @@
 
 #include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 #include "sw/device/lib/common.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
 
 #define PINMUX0_BASE_ADDR 0x40070000
 
-static void init_pinmux_reg(uint32_t reg, uint32_t size, uint32_t num_fields,
-                            uint32_t mask, uint32_t start_v) {
-  uint32_t reg_value = 0;
-  uint32_t reg_offset = 0;
+static dif_pinmux_t pinmux0;
 
-  for (uint32_t i = 0; i < NUM_MIO; ++i) {
-    reg_value |= ((i + start_v) & mask) << (size * (i % num_fields));
-    if ((i % num_fields) == (num_fields - 1)) {
-      REG32(reg + reg_offset) = reg_value;
-      reg_value = 0;
-      reg_offset += 4;
-    }
+void pinmux_init(void) {
+  mmio_region_t base_addr = mmio_region_from_addr(PINMUX0_BASE_ADDR);
+  if (dif_pinmux_init(base_addr, &pinmux0) != kDifPinmuxInitErrorNone) {
+    // TODO - print debug output.
+    return;
   }
 
-  REG32(reg + reg_offset) = reg_value;
-}
-
-void pinmux_init(void) {
-  // input: assign MIO0..MIO31 to GPIO0..GPIO31
-  init_pinmux_reg(PINMUX_PERIPH_INSEL0(0), PINMUX_PERIPH_INSEL0_IN1_OFFSET,
-                  32 / PINMUX_PERIPH_INSEL0_IN1_OFFSET,
-                  PINMUX_PERIPH_INSEL0_IN0_MASK,
-                  PINMUX_PERIPH_INSEL_IDX_OFFSET);
+  // Assign a MIO outputs to peripherl inputs (1:1, don't assign if peripheral
+  // input doesn't have a pair). Try to assign consecutively peripheral input
+  // 0,1,2... to MIO pad output 0,1,2... .
+  dif_pinmux_peripheral_in_t peripheral_in = kDifPinmuxPeripheralInGpio0;
+  dif_pinmux_insel_t mio_out_pad = kDifPinmuxInselPadFirst;
+  while (peripheral_in < DIF_PINMUX_PERIPHERAL_IN_NUM &&
+         mio_out_pad < DIF_PINMUX_INSEL_NUM) {
+    dif_pinmux_insel_set(&pinmux0, peripheral_in, mio_out_pad);
+    ++peripheral_in;
+    ++mio_out_pad;
+  }
 
-  // output: assign GPIO0..GPIO31 to MIO0..MIO31
-  init_pinmux_reg(PINMUX_MIO_OUTSEL0(0), PINMUX_MIO_OUTSEL0_OUT1_OFFSET,
-                  32 / PINMUX_MIO_OUTSEL0_OUT1_OFFSET,
-                  PINMUX_MIO_OUTSEL0_OUT0_MASK,
-                  PINMUX_PERIPH_OUTSEL_IDX_OFFSET);
+  // Assign a peripherl outputs to MIO inputs (1:1, don't assign if peripheral
+  // output doesn't have a pair). Try to assign consecutively peripheral output
+  // 0,1,2... to MIO pad input 0,1,2... .
+  dif_pinmux_mio_in_t mio_in_pad = kDifPinmuxMioInPadFirst;
+  dif_pinmux_outsel_t peripheral_out = kDifPinmuxOutselOutGpio0;
+  while (mio_in_pad < DIF_PINMUX_MIO_IN_NUM &&
+         peripheral_out < DIF_PINMUX_OUTSEL_NUM) {
+    dif_pinmux_outsel_set(&pinmux0, mio_in_pad, peripheral_out);
+    ++peripheral_out;
+    ++mio_in_pad;
+  }
```

Comment:
If we use separate functions for constants and peripheral/mio indices, we can perform these operations in a single loop.

Reply:
I didn't quite find any good way of doing it, and I think `pinmux.c` will probably be a relatively short-lived - eventually these operations will be called either directly in the `boot_rom`, or in a tock driver?